python/taichi/misc/gui.py

```diff
@@ -139,24 +139,57 @@ def show(self, file=None):
             self.core.screenshot(file)
         self.clear(self.background_color)
 
+    class EventFilter:
+        def __init__(self, *filter):
+            self.types = set()
+            self.keys = set()
+            self.combs = set()
+            for ent in filter:
+                if isinstance(ent, (tuple, list)):
+                    type, key = ent
+                    self.combs.add((type, key))
+                elif ent in [GUI.PRESS, GUI.RELEASE]:
+                    self.types.add(ent)
+                else:
+                    self.keys.add(ent)
+
+        def match(self, e):
+            if (e.type, e.key) in self.combs:
+                return True
+            if not self.keys or e.key in self.keys:
+                if not self.combs or e.key in self.combs:
```

Comment:
Do you mean to check `types`? Also, it appears to me that L159 and L160 has an OR relationship, so i'd still suggest  to write flattened if checks.

Reply:
> `types`

Nice catch! I couldn't imagine how my brain was fried to write such buggy code...

> Also

We want AND, see PR description, so that user can `get_event(ti.GUI.ESCAPE, ti.GUI.PRESS)` to filter only ESC press.

Reply:
Then it seems that we have a discrepancy here:

1. According to your `by comb` example

```py
filt = [(ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE)]
while not gui.get_event(*filt):  # PRESS and ESC  or  RELEASE and SPACE
```

The elements in `get_event()`'s `*filter` argument have an OR relationship.

Yet in the `by key + type` example:

```py
while not gui.get_event(ti.GUI.RELEASE, ti.GUI.ESCAPE):  # has REL and ESC event
    gui.show()
```

The elements in `get_event()`'s `*filters` now have an AND relationship.

---

What do you think if we make this consistent by doing this:

* elements in the same tuple have an AND relationship
* elements in the top-level `*filter` args always have an OR relationship

Then for the `by comb` example, these are equivalent:

```py
gui.get_event((ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE))
gui.get_event(*[(ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE)])
```

And for the `by key + type` example, it becomes:

```py
gui.get_event((ti.GUI.RELEASE, ti.GUI.ESCAPE))
```

Then, the following two become equivalent and mean either `ti.GUI.RELEASE` or `ti.GUI.ESCAPE`, although the second one is unnecessarily verbose:

```
gui.get_event(ti.GUI.RELEASE, ti.GUI.ESCAPE)
gui.get_event((ti.GUI.RELEASE, ), (ti.GUI.ESCAPE, ))
```
---

FYI, I should have mentioned this earlier, but please leave the work of "marking comments as resolved" to the reviewers. Otherwise it's easy that we might miss some comments in the previous reviews.

Reply:
Thanks, resolved now.